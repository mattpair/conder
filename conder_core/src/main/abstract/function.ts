import { AnySchemaInstance, AnyOpInstance, ow } from '../ops/index';
import { AnyNode, AnyRootNodeFromSet, BaseNodeDefs, RootNode } from "./IR";
import {MONGO_COMPILER, MONGO_GLOBAL_ABSTRACTION_REMOVAL} from './globals/mongo'
import { Transformer, Compiler } from "./compilers";


export type GlobalObject = {kind: "glob", name: string}

export type Manifest<F=FunctionDescription> = {
    globals: Map<string, GlobalObject>
    funcs: Map<string, F>
}

export type FunctionData<COMP=RootNode> = {
    readonly input: AnySchemaInstance[]
    readonly computation: COMP[]
}

export class FunctionDescription<COMP=RootNode> implements FunctionData<COMP>{
    public readonly input: AnySchemaInstance[]
    public readonly computation: COMP[]
    
    constructor(state: FunctionData<COMP>) {
        this.input = state.input
        this.computation = state.computation
    }

    public apply<NEW>(f: (c: COMP) => NEW[]): FunctionDescription<NEW> {
        return new FunctionDescription({
            input: this.input,
            computation: this.computation.flatMap(f)
        })
    }
}


export function toOps(funcs: Map<string, FunctionDescription>, override:  Compiler<RootNode> | undefined=undefined): Map<string, AnyOpInstance[]> {
    const ret: Map<string, AnyOpInstance[]> = new Map()
    const compiler: Compiler<RootNode> = override ? override : MONGO_GLOBAL_ABSTRACTION_REMOVAL.then(MONGO_COMPILER)
    // const computationLookup: Record<string, RootNode[]> = {}

    // ops.push(...compiler.run(funcs[k].computation))
    funcs.forEach((func, func_name) => {
        const ops: AnyOpInstance[] = [
            ow.assertHeapLen(func.input.length)
        ]
        func.input.forEach((schema, index) => {
            ops.push(
                ow.enforceSchemaInstanceOnHeap({heap_pos: index, schema}),
                ow.conditonallySkipXops(1),
                ow.raiseError("invalid input")
            )
        })
    
    
        ret.set(func_name, ops)
    })

    const compiled = compiler.run(funcs)
    funcs.forEach((func, func_name) => {
        ret.set(func_name, [...ret.get(func_name), ...compiled.get(func_name).computation])
    })
    return ret
}


export const OPSIFY_MANIFEST = new Transformer<
    Manifest, 
    Manifest<AnyOpInstance[]>>((man) => {
        const funcs = toOps(man.funcs)
        return {
            funcs,
            globals: man.globals
        }
    })