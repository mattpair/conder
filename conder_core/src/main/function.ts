import { AnySchemaInstance, AnyOpInstance, ow } from "conder_kernel";
import { AnyNode , compile} from "./DAG";


export type FunctionDescription = {
    input: AnySchemaInstance[]
    computation: AnyNode[]
}

export function toOps(func: FunctionDescription): AnyOpInstance[] {
    const ops: AnyOpInstance[] = [
        ow.assertHeapLen(func.input.length)
    ]
    func.input.forEach((schema, index) => {
        ops.push(
            ow.enforceSchemaInstanceOnHeap({heap_pos: index, schema}),
            ow.conditionalOpOffset(2),
            ow.raiseError("invalid input")
        )
    })
    ops.push(...compile(...func.computation))
    return ops
}