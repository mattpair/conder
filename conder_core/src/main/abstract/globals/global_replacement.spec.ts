import { MONGO_GLOBAL_ABSTRACTION_REMOVAL, AnyNode, AbstractNodes } from "../../../../index"
import { FunctionDescription } from "../function"


describe("mongo", () => {

    function replaceTest(original: Exclude<AnyNode, AbstractNodes>): jest.ProvidesCallback {
        return (cb) => {
            const map = new Map<string, FunctionDescription>().set("func", {   
                // Typically, a root node is required so you express a meaningful computation.
                // However, it would make tests more verbose here and the abstraction remover
                // is capable of handling any type of node.
                computation: [original] as any, 
                input: []
            })
            expect(MONGO_GLOBAL_ABSTRACTION_REMOVAL.run(map)).toMatchSnapshot()
            cb()
        }
    }
    it("get field with mongo specific op", replaceTest({
            kind: "GetField",
            target: {kind: "GlobalObject", name: "global"},
            field_name: [{kind: "String", value: "field"}]
        })
    )

    it("can replace existence checking", replaceTest({   
            kind: "FieldExists",
            value: {kind: "GlobalObject", name: "glob"},
            field: {kind: "String", value: "maybe"}
        })
    )

    it("can replace SetField updates", replaceTest({
        kind: "Update",
        target: {kind: "GlobalObject", name: "gg"},
        operation: {kind: "SetField", value: {kind: "String", value: 'some val'}, field_name: [{kind: "Saved", index: 12}]}
    }))
})