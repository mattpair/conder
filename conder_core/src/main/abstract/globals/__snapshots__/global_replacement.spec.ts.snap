// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mongo can replace SetField updates 1`] = `
Map {
  "func" => FunctionDescription {
    "computation": Array [
      Object {
        "key": Array [
          Object {
            "index": 12,
            "kind": "Saved",
          },
        ],
        "kind": "SetKeyOnObject",
        "obj": "gg",
        "value": Object {
          "kind": "String",
          "value": "some val",
        },
      },
    ],
    "input": Array [],
  },
}
`;

exports[`mongo can replace existence checking 1`] = `
Map {
  "func" => FunctionDescription {
    "computation": Array [
      Object {
        "key": Object {
          "kind": "String",
          "value": "maybe",
        },
        "kind": "keyExists",
        "obj": "glob",
      },
    ],
    "input": Array [],
  },
}
`;

exports[`mongo get field with mongo specific op 1`] = `
Map {
  "func" => FunctionDescription {
    "computation": Array [
      Object {
        "key": Array [
          Object {
            "kind": "String",
            "value": "field",
          },
        ],
        "kind": "GetKeyFromObject",
        "obj": "global",
      },
    ],
    "input": Array [],
  },
}
`;
